core/src/test/scala/org/apache/spark/scheduler/BlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,503 @@
+    def failOneTaskInTaskSet(exec: String): Unit = {
+      val taskSet = FakeTask.createTaskSet(1, stageId, 0)
+      val tsm = new TaskSetManager(scheduler, Some(tracker), taskSet, 1, clock)
+      tsm.updateBlacklistForFailedTask("host-" + exec, exec, 0)
+      tracker.updateBlacklistForSuccessfulTaskSet(stageId, 0, tsm.execToFailures)
+      stageId += 1
+    }
+    failOneTaskInTaskSet("1")
+    // We have one sporadic failure on exec 2 -- it doesn't lead to an exec blacklist.
+    failOneTaskInTaskSet("2")
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set())
+    assert(tracker.nextExpiryTime === Long.MaxValue)
+
+    // We advance the clock past the expiry time.
+    clock.advance(tracker.BLACKLIST_TIMEOUT_MILLIS + 1)
+    val t0 = clock.getTimeMillis()
+    tracker.applyBlacklistTimeout()
+    assert(tracker.nextExpiryTime === Long.MaxValue)
+    failOneTaskInTaskSet("1")
+
+    // Because we went past the expiry time, nothing should have been blacklisted.
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set())
+
+    // Now we add one more failure, within the timeout, and it should be counted.
+    clock.setTime(t0 + tracker.BLACKLIST_TIMEOUT_MILLIS)
+    val t1 = clock.getTimeMillis()
+    failOneTaskInTaskSet("1")
+    tracker.applyBlacklistTimeout()
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set("1"))
+    assert(tracker.nextExpiryTime === t1 + tracker.BLACKLIST_TIMEOUT_MILLIS)
+
+    // Fail a second executor, and go over its expiry as well.
+    clock.setTime(t1 + tracker.BLACKLIST_TIMEOUT_MILLIS)
+    val t2 = clock.getTimeMillis()
+    failOneTaskInTaskSet("3")
+    failOneTaskInTaskSet("3")
+    tracker.applyBlacklistTimeout()
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set("1", "3"))
+    assert(tracker.nextExpiryTime === t1 + tracker.BLACKLIST_TIMEOUT_MILLIS)
+
+
+    clock.setTime(t1 + tracker.BLACKLIST_TIMEOUT_MILLIS + 1)
+    tracker.applyBlacklistTimeout()
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set("3"))
+    assert(tracker.nextExpiryTime === t2 + tracker.BLACKLIST_TIMEOUT_MILLIS)
+
+
+    // Make sure that we update correctly when we go from having blacklisted executors to
+    // just having tasks with timeouts.
+    clock.setTime(t2 + tracker.BLACKLIST_TIMEOUT_MILLIS)
+    val t3 = clock.getTimeMillis()
+    failOneTaskInTaskSet("4")
+    tracker.applyBlacklistTimeout()
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set("3"))
+    assert(tracker.nextExpiryTime === t2 + tracker.BLACKLIST_TIMEOUT_MILLIS)
+
+    clock.setTime(t2 + tracker.BLACKLIST_TIMEOUT_MILLIS + 1)
+    val t4 = clock.getTimeMillis()
+    tracker.applyBlacklistTimeout()
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set())
+    // we've got one task failure still, but we don't bother setting nextExpiryTime to it, to
+    // avoid wasting time checking for expiry of individual task failures.
+    assert(tracker.nextExpiryTime === Long.MaxValue)
```

Comment:
yes in a way, but it seems worth checking again, since the earlier check is more of an initial condition + no update, while this is checking that it is getting reset correctly.